pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -106,14 +106,24 @@ func NewGarbageCollector(
 		sharedInformers:  sharedInformers,
 		ignoredResources: ignoredResources,
 	}
-	if err := gb.monitorsForResources(deletableResources); err != nil {
-		return nil, err
+	if err := gb.syncMonitors(deletableResources); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to sync all monitors: %v", err))
 	}
 	gc.dependencyGraphBuilder = gb
 
 	return gc, nil
 }
 
+// SyncResourceMonitors starts or stops resource monitors as needed to ensure
+// that all (and only) those resources present in the map are monitored.
+func (gc *GarbageCollector) SyncResourceMonitors(deletableResources map[schema.GroupVersionResource]struct{}) error {
```

Comment:
We need to disable all the workers, call this function, then wait for dependencyGraphBuilder.HasSynced(), then enable the workers again. Otherwise we will hit this issue: https://github.com/kubernetes/kubernetes/issues/28713#issuecomment-234826276